src/server.c

```diff
@@ -1051,7 +1051,17 @@ void removeClientFromMemUsageBucket(client *c, int allow_eviction) {
  * returns 1 if client eviction for this client is allowed, 0 otherwise.
  */
 int updateClientMemUsageAndBucket(client *c) {
-    serverAssert(pthread_equal(pthread_self(), server.main_thread_id) && c->conn);
+    /* The unlikely case this function was called from a thread different
+     * than the main one is a module call from a spawned thread. This is safe
+     * since this call must have been made after calling
+     * RedisModule_ThreadSafeContextLock i.e the module is holding the GIL. In
+     * that special case we assert that at least the updated client's
+     * running_tid is the main thread. The true main thread is allowed to call
+     * this function on clients handled by IO-threads as it makes sure the
+     * IO-threads are paused, f.e see cleintsCron() and evictClients(). */
+    serverAssert((pthread_equal(pthread_self(), server.main_thread_id) ||
+                               c->running_tid == IOTHREAD_MAIN_THREAD_ID) &&
+                 c->conn);
```

Comment:
@ShooterIT It seems that the indentation after this modification has become even stranger. The discomfort lies in the fact that c->conn occupies a separate line.
what about
```suggestion
    serverAssert((pthread_equal(pthread_self(), server.main_thread_id) ||
                  c->running_tid == IOTHREAD_MAIN_THREAD_ID) && c->conn);
```

Reply:
I agree it is a bit ugly to put the second condition at a single line, just want the second condition more clear.

i am OK for your changes.

Reply:
Looking at big conditions like this one in the source code I would suggest the following
```c
serverAssert(c->conn &&
             (pthread_equal(pthread_self(), server.main_thread_id) ||
              c->running_tid == IOTHREAD_MAIN_THREAD_ID));
```

Reply:
also LGTM